### PR TITLE
[New package] Add HIP_jll 4.0.0

### PR DESCRIPTION
**Pull request by jpsamaroo**
WIP until I can get rocBLAS, rocFFT, and rocRAND built and working.

TODO:
- [x] Patch hipcc to find C++ headers, libs, and C runtime

### CODE REVIEW

**Comment by jpsamaroo** [comment]
I'm able to use this with AMDGPU locally, and also built rocRAND with it (which I'll post after this is merged). In my eyes, this is good to go.

File: H/HIP/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+atomic_patch -p1 "${WORKSPACE}/srcdir/patches/hipcc-bb-paths.patch"
+
+# use rocclr instead of hcc as HIP_PLATFORM
+atomic_patch -p1 "${WORKSPACE}/srcdir/patches/rocclr-not-hcc-platform.patch"
+#sed -e "s:\$HIP_PLATFORM eq \"hcc\" and \$HIP_COMPILER eq \"clang\":\$HIP_PLATFORM eq \"rocclr\" and \$HIP_COMPILER eq \"clang\":" -i bin/hipcc
+
+apk add coreutils dateutils
+
+mkdir build && cd build
+export HIP_CLANG_PATH=$WORKSPACE/destdir/tools
```
Comment (giordano) [inline H/HIP/build_tarballs.jl:33]:
```suggestion
export HIP_CLANG_PATH=${prefix}/tools
```

File: H/HIP/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN%.*}_clang.cmake \
+      -DLLVM_DIR="${prefix}/lib/cmake/llvm" \
+      -DClang_DIR="${prefix}/lib/cmake/clang" \
+      -DLLD_DIR="${prefix}/lib/cmake/ldd" \
+      ..
+make -j${nproc}
+make install
+
+# link to .hipVersion so clang can find it
+ln -s $WORKSPACE/destdir/hip/bin/.hipVersion $WORKSPACE/destdir/bin/
```
Comment (giordano) [inline H/HIP/build_tarballs.jl:51]:
```suggestion
ln -s ../hip/bin/.hipVersion ${bindir}/.hipVersion
```

File: H/HIP/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      -DCMAKE_BUILD_TYPE=Release \
+      -DHIP_COMPILER=clang \
+      -DHIP_PLATFORM=rocclr \
+      -DHIP_RUNTIME=ROCclr \
+      -DROCM_PATH=${prefix} \
+      -DHSA_PATH=${prefix} \
+      -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN%.*}_clang.cmake \
+      -DLLVM_DIR="${prefix}/lib/cmake/llvm" \
+      -DClang_DIR="${prefix}/lib/cmake/clang" \
+      -DLLD_DIR="${prefix}/lib/cmake/ldd" \
```
Comment (giordano) [inline H/HIP/build_tarballs.jl:45]:
```suggestion
      -DLLVM_DIR="${libdir}/cmake/llvm" \
      -DClang_DIR="${libdir}/cmake/clang" \
      -DLLD_DIR="${libdir}/cmake/ldd" \
```

**Pull request merged**